### ISSUE TITLE
What is the difference between treemap and opt?

### ISSUE THREAD

**Issue by Waterkin** [deepmind/optax#339]
**_For params 'p', use sample x and y to update it, what's the difference between following codes?_**
`grads = jax.grad(loss)(p, x, y)`
`updates, state = opt.update(grads, state)`
`p = optax.apply_updates(p, updates)`

`grads = jax.grad(loss)(p, x, y)`
`inner_sgd_fn = lambda g, state: (state - alpha*g)`
`p = jax.tree_util.tree_multimap(inner_sgd_fn, grads, p)`

**Comment by n2cholas** [comment]
If your `opt` is SGD, then those two code snippets should do the same thing. However, if you are using a different optimization method (e.g. Momentum, Adam, RMSProp, etc), then there is additional state (e.g. momentum terms) that Optax will help you manage. This quickstart notebook might illustrate this better: https://github.com/deepmind/optax/blob/master/examples/quick_start.ipynb

Does that make sense?
